### PR TITLE
CFC-83: Add Caliber Limits

### PR DESCRIPTION
**Pull request by sarahsturgeon**
Tested locally, seems fine, but I would encourage you to double-check.

~~**This relies on an ACF-3-Missiles PR**: https://github.com/TwistedTail/ACF-3-Missiles/pull/83~~

image

### CODE REVIEW

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 1]
```diff
@@ -0,0 +1,152 @@
+util.AddNetworkString( "ACFBalance_CaliberLimitAlert" )
+
+local WHITE = Color( 255, 255, 255 )
+local LIGHTRED = Color( 175, 25, 25 )
+local LIGHTGREY = Color( 200, 200, 200 )
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: Should be SCREAMING_SNAKE_CASE, not SCREAMINGCASE

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+local maxCaliber = 280
+local caliberLimit = 450
+local missileCaliberScale = 0.5
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: Constants should be SCREAMING_SNAKE_CASE.
Might be handy to convert these to convars as well for easier tweaking and testing, although it's not required.

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    local caliber = data.Caliber or 40
+    if what == "acf_missile" then
+        caliber = caliber * missileCaliberScale
+    end
+
+    return caliber
+end
+
+local guns = "ACFBalance_Guns"
+local calibers = "ACFBalance_TotalCaliber"
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: These variable names are misleading and are very easy to accidentally shadow.

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            start, finish = YELLOW, RED
+        end
+
+        return LerpVector( range, start, finish ):ToColor()
+    end
+end
+
+local queueCaliberMessage
+do
+    local timerPrefix = "ACFBalance_CaliberLimitAlert_"
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: Constants should be SCREAMING_SNAKE_CASE.

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    local function sendCaliberMessage( ply, prefix )
+        prefix = prefix or {}
+
+        if not ply and ply:IsValid() then return end
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: RERO; fix logic.
```suggestion
        if not ply or not ply:IsValid() then return end

        prefix = prefix or {}
```

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+        -- Just in case one was already queued somehow
+        timer.Remove( timerPrefix .. ply:SteamID64() )
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: Maybe change this to a local function` getTimerName( ply )` or `getQueueTimerName( ply )` to ensure consistency throughout. Would also drop the need for a constant variable to hold the string.

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        timer.Create( timerPrefix .. ply:SteamID64(), delay, 1, function()
+            sendCaliberMessage( ply, prefix )
+        end )
+    end
+
+    hook.Add( "PlayerSay", "ACFBalance_CaliberLimit", function( ply, text )
+        if text ~= "!calibers" then return end
+
+        queueCaliberMessage( ply )
+        return ""
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: Style requires a newline before returns.
```suggestion
        queueCaliberMessage( ply )

        return ""
```

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    local hasCalibers = {
+        acf_gun = true,
+        acf_missile = true
+    }
+
+    hook.Add( "ACF_OnEntitySpawn", "ACFBalance_CaliberLimit", function( what, ent, entData )
+        if not hasCalibers[what] then return end
+
+        local ply = ent.Owner
+        local plyTable = ply:GetTable()
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua:168]: Is it possible for ACF entities to be spawned without having a player assigned as the owner?

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        local gunCaliber = getCaliber( what, entData )
+
+        plyTable[guns][ent] = gunCaliber
+        ent:CallOnRemove( "ACFBalance_CaliberLimit", function()
+            if not ply and ply:IsValid() then return end
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]: Fix logic.
```suggestion
            if not ply or not ply:IsValid() then return end
```
Could also drop the `not ply or` if a validity check is done prior, ensuring that it's non-nil.

**Review by legokidlogan — DISMISSED** [review]

File: lua/autorun/server/sv_acf_caliber_limits.lua (hunk 2 above)
Comment (sarahsturgeon) [reply lua/autorun/server/sv_acf_caliber_limits.lua]: I'm kind of getting 👎 on screaming snake for constants.
It makes the code long when you go to actually use them...

Comment (sarahsturgeon) [reply lua/autorun/server/sv_acf_caliber_limits.lua]: Caliber limit should be a convar, agreed

File: lua/autorun/server/sv_acf_caliber_limits.lua (hunk 8 above)
Comment (sarahsturgeon) [reply lua/autorun/server/sv_acf_caliber_limits.lua:168]: We'll find out 😎

File: lua/autorun/server/sv_acf_caliber_limits.lua (hunk 9 above)
Comment (sarahsturgeon) [reply lua/autorun/server/sv_acf_caliber_limits.lua]: Gotta be done here, their props could be removed after they leave

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 10]
```diff
@@ -51,35 +96,38 @@ end
 
 local queueCaliberMessage
 do
-    local timerPrefix = "ACFBalance_CaliberLimitAlert_"
+    local timerName = function( ply )
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]:
```suggestion
    local function getTimerName( ply )
```

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 11]
```diff
@@ -51,35 +96,38 @@ end
...
     local function sendCaliberMessage( ply, prefix )
         prefix = prefix or {}
 
-        if not ply and ply:IsValid() then return end
+        if not ply then return end
+        if not ply:IsValid() then return end
 
         -- Just in case one was already queued somehow
-        timer.Remove( timerPrefix .. ply:SteamID64() )
+        timer.Remove( timerName( ply ) )
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]:
```suggestion
        timer.Remove( getTimerName( ply ) )
```

File: lua/autorun/server/sv_acf_caliber_limits.lua [hunk 12]
```diff
@@ -93,7 +141,7 @@ do
     queueCaliberMessage = function( ply, prefix, delay )
         delay = delay or 0.15
 
-        timer.Create( timerPrefix .. ply:SteamID64(), delay, 1, function()
+        timer.Create( timerName( ply ), delay, 1, function()
```
Comment (legokidlogan) [inline lua/autorun/server/sv_acf_caliber_limits.lua]:
```suggestion
        timer.Create( getTimerName( ply ), delay, 1, function()
```

**Review by legokidlogan — CHANGES_REQUESTED** [review]

**Review by legokidlogan — APPROVED** [review]

**Pull request merged**